### PR TITLE
Make styled property storage typed

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Previously, when a value was set on a styled property, that property was boxed and added to the `ValueStore`. This PR refactors the value store and the API of `AvaloniaObject` to store values without boxing.

`IAvaloniaObject` has been changed to have two overloads of each get/set/bind/clear method:

- an overload which takes `StyledPropertyBase<T>`
- an overload which takes `DirectPropertyBase<T>`

In addition, the `IObservable<>` passed to `Bind` has changed to an `IObservable<BindingValue<T>>`: `BindingValue<T>` is a struct which allows representation of extra state in a typed manner:

- A simple value
- An unset value
- Do nothing
- Binding errors (with optional fallback value)
- Data validation errors (with optional fallback value)

Because the get/set/bind/clear methods are now typed on the value `T`, the value store can now store values without boxing.

Untyped `AvaloniaProperty` access is still allowed and calls are routed through the `AvaloniaProperty` itself to the correct generic method.

Bindings which produce an `IObservable<T>` are wrapped in an adapter to convert them to `IObservable<BindingValue<T>>`.

## Benchmarks

`master`:

```
|                                Method |     Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |---------:|----------:|----------:|--------:|------:|------:|----------:|
|           Set_Int_Property_LocalValue | 19.93 us | 0.0541 us | 0.0452 us | 11.1084 |     - |     - |  10.45 KB |
|  Set_Int_Property_Multiple_Priorities | 72.47 us | 0.2524 us | 0.2361 us | 19.8975 |     - |     - |  18.73 KB |
|      Set_Int_Property_TemplatedParent | 27.86 us | 0.1242 us | 0.1101 us |  9.0637 |     - |     - |   8.52 KB |
|          Bind_Int_Property_LocalValue | 18.03 us | 0.2491 us | 0.2080 us |  9.3689 |     - |     - |    8.8 KB |
| Bind_Int_Property_Multiple_Priorities | 39.44 us | 0.3584 us | 0.3352 us | 21.9116 |     - |     - |  20.63 KB |
```

This PR:

```
|                                Method |       Mean |     Error |    StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |-----------:|----------:|----------:|--------:|------:|------:|----------:|
|           Set_Int_Property_LocalValue |  14.793 us | 0.0612 us | 0.0511 us |  0.2594 |     - |     - |     256 B |
|  Set_Int_Property_Multiple_Priorities | 217.471 us | 0.8287 us | 0.6470 us | 25.6348 |     - |     - |   24696 B |
|      Set_Int_Property_TemplatedParent |  34.932 us | 0.4928 us | 0.4610 us |  6.7139 |     - |     - |    6504 B |
|          Bind_Int_Property_LocalValue |   9.989 us | 0.0843 us | 0.0788 us |  0.4120 |     - |     - |     400 B |
| Bind_Int_Property_Multiple_Priorities |  41.421 us | 0.1188 us | 0.1053 us |  1.5259 |     - |     - |    1512 B |
```

*Note that the `Set_Int_Property_Multiple_Priorities` and `Set_Int_Property_TemplatedParent ` tests are now slower/allocate more due to the change in behavior listed under "Breaking Changes" -> "Values". However despite the fact they're storing multiple values instead of a single value, the total allocated memory is still a fraction of what it was before*

## Not addressed in this PR

- All of our current infrastructure that produces bindings are still typed as `IObservable<object>` meaning that they have to be wrapped. This means that you won't see much perf improvement from this PR alone. Changes to fix this will come in later PRs.
- ~~The stackoverflow problems described in #855 and #824 are not addressed. We previously used `DeferedSetter` to get around these problems, but that had a non-zero performance impact. The fix for these issues will come in a separate PR to make reviewing of just that part more easy.~~ Merged into this PR by #3258 
- More of the `AvaloniaObject.GetValue`/`SetValue`/`Bind` methods can be moved to extension methods, but that will require a change to the XAML compiler

## Checklist

- [x] Added unit tests (if possible)?

### Properties

- If your readonly property fields are declared as of type `AvaloniaProperty<T>` then you should change them to `StyledProperty<>` or `DirectProperty<,>` fields
- The signature for `OnPropertyChanged` has changed and is now a generic method in order to avoid allocation and boxing. To cast the `oldValue` and `newValue` parameters to a concrete type, use `.ValueOrDefault<T>()` on `Optional<T>` and `BindingValue<T>`

### Values

Previously all values set via `SetValue` were only active until a binding of the same or higher priority produced a value, in which case the `SetValue` value was lost.

This meant that style setters with needed to create a binding with a single value, otherwise the setter value would be lost when e.g. a setter with `StyleTrigger` binding was activated.

This PR changes that behavior: `LocalValue` priority values now work as before but other priority values are stored alongside bindings in the entries for the priority value and as such will not be overwritten by binding values. This will allow more efficient style values in a future PR as single values will no longer have to be created as single-value non-completing bindings

### Binding

- Many `AvaloniaObject.Bind` overloads have been moved to be extension methods, so you may have to add `this.` to your `Bind()` call when binding to `this`

### Validation

~~Styled property value validation and coercion was a major performance bottleneck, because it imposed a metadata lookup on _every_ styled property set and binding update, regardless of whether validation was enabled for the property (and for 99% of properties, it isn't)~~

~~In WPF/UWP value validation is defined outside of metadata, so we could bring validation back in a later PR in this form if needed. However even in WPF/UWP coercion imposes the overhead of a metadata lookup so it's unlikely we'd bring this back. The problem is less of an issue in Avalonia due to direct properties however.~~

~~#3287 implements WPF-style property validation and coercion and targets this branch. I've kept it as a separate PR for now to ease review. #3287 should be merged into this PR before this PR is merged.~~ #3287 merged

Avalonia properties now take a separate validation and coercion callback the same as WPF. The validation callback cannot be overridden, though the coercion callback can.

Depends on #3287

### CODE REVIEW

File: src/Avalonia.Base/ValueStore.cs
Comment (MarchingCube) [inline src/Avalonia.Base/ValueStore.cs]: This is causing boxing of all non-reference types values that have local priority. Is local value priority not important from performance perspective?

File: src/Avalonia.Controls/Button.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Button.cs]: This change here is a bit... random.

**Review by MarchingCube — COMMENTED** [review]
Glanced over most of your changes. I think it would be good to figure out if we can address `LocalValue` setter performance. After cobbling together a quick hack to store typed local values I was able to improve memory usage further and improve perf by like 10-15%.

|                      Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| Set_Int_Property_LocalValue | 11.92 us | 0.2315 us | 0.2573 us | 0.0610 |     - |     - |     256 B |

**Comment by mstr2** [comment]
Will this remove value coercion for `StyledProperty` altogether?

File: src/Avalonia.Base/Data/Optional.cs
Comment (grokys) [reply src/Avalonia.Base/Data/Optional.cs]: Would it be best to declare this as non-readonly then?

File: src/Avalonia.Base/Diagnostics/AvaloniaObjectExtensions.cs
Comment (grokys) [reply src/Avalonia.Base/Diagnostics/AvaloniaObjectExtensions.cs]: This is the last piece of functionality I have to implement before taking the PR out of draft state.

File: src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs
Comment (grokys) [reply src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs]: I do have some ideas for a better way to do this in a future PR, but I'd like to understand what you're thinking here. What property changed observable itself?

File: src/Avalonia.Base/ValueStore.cs
Comment (grokys) [reply src/Avalonia.Base/ValueStore.cs]: This was actually my attempt at an optimization, ported from the old `ValueStore`. The idea being that the most common case is a single local value: rather than creating a new `ConstantValueEntry` to wrap the value, just box the value and store it directly in the `AvaloniaPropertyValueStore`. In the absence discriminated unions in C#, `AvaloniaPropertyValueStore`'s value type must be heap allocated anyway.

I'd be interested in seeing the quick hack for storing typed local values you mentioned.

File: src/Avalonia.Controls/Button.cs
Comment (grokys) [reply src/Avalonia.Controls/Button.cs]: I seem to have swapped the order of the `OnPointerCaptureLost`/`UpdateDataValidation` methods, I'm not sure though why git's diff is marking this as added, it's unchanged from diff on the left-hand-side.

**Comment by grokys** [comment]
> Will this remove value coercion for StyledProperty altogether?

It will, yes. Is this a problem for you? We may need to design a replacement before we commit this change. As I mentioned, the current way coercion works slows down every single property set (including values from bindings), so it'd probably have a different API/semantics.

**Comment by mstr2** [comment]
> It will, yes. Is this a problem for you? We may need to design a replacement before we commit this change. As I mentioned, the current way coercion works slows down every single property set (including values from bindings), so it'd probably have a different API/semantics.

Yes, I’m using it for data consistency. It seems to me that it should be pretty easy to optimize the case where a property doesn’t use validation on any type by having a flag on the property that specifies whether any of its metadata has a validation function.

File: src/Avalonia.Base/Data/Optional.cs
Comment (MarchingCube) [reply src/Avalonia.Base/Data/Optional.cs]: I think so, otherwise we can cause a lot of extra copies.  Or we make this a `readonly` struct as well.

File: src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs
Comment (MarchingCube) [reply src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs]: I actually checked this later and property observable is actually implemented using this class :) My derp. I meant `GetObservable` method.

File: src/Avalonia.Base/ValueStore.cs
Comment (MarchingCube) [reply src/Avalonia.Base/ValueStore.cs]: Sure, when I come back from work I can share it. Hopefully I still have it around :) But I think the gist of it was implementing `IValue` and reusing that instead of boxing.

File: src/Avalonia.Base/Data/Optional.cs
Comment (grokys) [reply src/Avalonia.Base/Data/Optional.cs]: Uh... yeah that would make more sense :)

**Comment by mstr2** [comment]
Since this is a breaking API change anyway: did you consider moving the static registration methods to their respective subclass? I think it's a bit cleaner to write:

```C#
StyledProperty<double> FooProperty = StyledProperty.Register<...>(...);
DirectProperty<double> BarProperty = DirectProperty.Register<...>(...);
```

File: src/Avalonia.Base/ValueStore.cs
Comment (MarchingCube) [reply src/Avalonia.Base/ValueStore.cs]: Delivering hack payload: https://github.com/MarchingCube/Avalonia/commit/08318164a73e64661eb177b011981ff20ba50dc8.

Maybe I missed something but it allocates less and runs faster on my PC.

**Comment by grokys** [comment]
> Since this is a breaking API change anyway: did you consider moving the static registration methods to their respective subclass? I think it's a bit cleaner to write:

We could move them and if necessary keep the existing deprecated methods, but this should be done in a separate PR. However, I'm not completely sure that it's a change worth making at this point, when everyone's code is using the existing methods...

File: src/Avalonia.Base/ValueStore.cs
Comment (grokys) [reply src/Avalonia.Base/ValueStore.cs]: Yep, you're right! I implemented something based on your hack and `Set_Int_Property_LocalValue`  allocated memory goes from 2632B -> 256B and faster to boot. Nice ;)

**Comment by grokys** [comment]
@mstr2 regarding coercion/validation, any thoughts on the current API as compared to the WPF/UWP API where validation is a "static" property and only coercion receives a reference to the control?

https://docs.microsoft.com/en-us/dotnet/framework/wpf/advanced/dependency-property-callbacks-and-validation

Not sure whether to change our API to match WPF/UWP or not.

File: src/Avalonia.Base/Utilities/AvaloniaPropertyValueStore.cs
Comment (grokys) [reply src/Avalonia.Base/Utilities/AvaloniaPropertyValueStore.cs]: Yeah we should implement something like `FrugalMap`, but in a separate PR I think.

**Comment by mstr2** [comment]
> @mstr2 regarding coercion/validation, any thoughts on the current API as compared to the WPF/UWP API where validation is a "static" callback and only coercion receives a reference to the control?
> Not sure whether to change our API to match WPF/UWP or not.

From a very high-level perspective, it is my understanding that validation and coercion in WPF address two slightly different things.

Class-level validation is used to enforce that all valid objects be a class (as in set theory). More precisely, all valid objects share a common property, and an object that does not have that property can never be a valid object.

Instance-level coercion, on the other hand, is a constrained range of valid objects. This range can depend on many things and might change at any time.

While it might be elegant in theory, actual production code will often not make that distinction in practice. I understand that Avalonia is currently using a simpler approach, where the distinction between these two aspects is done by throwing an exception (object does not have a required property) or returning a coerced version of the object (object is valid, but out of range).

All other things being equal, I will always prefer the WPF behavior for the simple reason that matching this behavior will mean that, when faced with a coding problem, someone looking up solutions on the internet will find it more likely that WPF solutions will work in Avalonia.

**Comment by grokys** [comment]
> All other things being equal, I will always prefer the WPF behavior for the simple reason that matching this behavior will mean that, when faced with a coding problem, someone looking up solutions on the internet will find it more likely that WPF solutions will work in Avalonia.

I agree. I will re-implement validation/coercion based on the WPF/UWP API then.

File: src/Avalonia.Base/Data/Optional.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/Optional.cs]: Perf issue that I spotted when profiling your code - you are double checking `HasValue` in many places. This function ends up checking `HasValue` twice and bloats this function so it won't be inlined in many cases.

**Review by MarchingCube — CHANGES_REQUESTED** [review]

File: src/Avalonia.Base/Data/Optional.cs
Comment (MarchingCube) [reply src/Avalonia.Base/Data/Optional.cs]: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Nullable.cs

Comment (grokys) [reply src/Avalonia.Base/Data/Optional.cs]: RE: double checking `HasValue`, can you explain more? Do you mean that it's checked in `ValueOrDefault` and then in `Value`? The `Nullable` code that you linked to appears to do the same, is it different because it checks a field and not a property?

Also I should rename this method `GetValueOrDefault` to match `Nullable`.

Comment (MarchingCube) [reply src/Avalonia.Base/Data/Optional.cs]: Nullable code does it a bit differently:
```csharp
public T GetValueOrDefault(T defaultValue) =>
            hasValue ? value : defaultValue;
```

In most of the code it tries to avoid using `Value` if we know that it is set (by checking `hasValue`), otherwise like in your code we end up doing:
- check if value is set
    - it is set
        - check if value is set (entering potential EH scope)
        - return value
    - it is not set
        - return default value

Comment (grokys) [reply src/Avalonia.Base/Data/Optional.cs]: Ah yeah, apologies, I missed the fact that `Nullable` accesses the field directly in `GetValueOrDefault`.

Comment (grokys) [reply src/Avalonia.Base/Data/Optional.cs]: @MarchingCube hopefully fixed this in recent commits.

**Comment by grokys** [comment]
Now that #3255 re-implements validation/coercion, this is ready for review.

File: src/Avalonia.Base/Data/BindingValue.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/BindingValue.cs]: Compiler warning for hidden overload (function above will be picked over this one when no argument has been specified so this `= default` is not needed).

File: src/Avalonia.Base/Data/BindingValue.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/BindingValue.cs]: Compiler warning for hidden overload (function above will be picked over this one when no argument has been specified so this `= default` is not needed).

File: src/Avalonia.Base/Data/BindingValue.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/BindingValue.cs]: Typo in `Invalid`. English is preferred over Italian :)

File: src/Avalonia.Base/Data/Optional.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/Optional.cs]: No point in having default value if you have an overload taking no parameters.

File: src/Avalonia.Base/PropertyStore/BindingEntry.cs
Comment (MarchingCube) [inline src/Avalonia.Base/PropertyStore/BindingEntry.cs]: Unused?

**Review by MarchingCube — COMMENTED** [review]
I've found a few small issues but I think generally your changes look fine.

I did a quick profile of ControlCatalog and things are generally looking good. One can clearly see overhead from styling now and hopefully with these changes we could reduce it as well later on.

Feedback that can be resolved later:
- `Visual.AffectsRender` shows up when profiling since local `Invalidate` function will box property values (`Opacity` being the prime one), maybe with new typed APIs we could avoid this.

I will get back to this review soon to do a second pass over it.

File: src/Avalonia.Controls.DataGrid/DataGrid.cs
Comment (grokys) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: They're restored in #3287 which should be merged into this PR before merging (I did mention this in the description but I linked to the wrong PR - fixed that now)

Comment (MarchingCube) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: Ah, right that circular PR link confused me a bit.

Comment (grokys) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: Yeah sorry about that!

File: src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Reactive/AvaloniaPropertyObservable.cs]: I got curious about this code since it will box value types multiple times. I've added a simple check if `e.Property` is `AvaloniaProperty<T>` and in ControlCatalog 60% of times it is. This allows this code to avoid all boxing and just use typed values. We still need a codepath for `AvaloniaPropertyObservable<object>` which is used by bindings code (hopefully to be typed as well eventually).

File: src/Avalonia.Base/Reactive/BindingValueExtensions.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Reactive/BindingValueExtensions.cs]: I think those both would need to return `IObservable<object?>` to satisfy nullability requirements and get rid of the warnings.

File: src/Avalonia.Base/IAvaloniaObject.cs
Comment (MarchingCube) [inline src/Avalonia.Base/IAvaloniaObject.cs]: Typo in `receive` :)

File: src/Avalonia.Base/StyledPropertyMetadata`1.cs
Comment (MarchingCube) [inline src/Avalonia.Base/StyledPropertyMetadata`1.cs]: You can also remove `Cast` function since it is not being used anymore.

**Review by MarchingCube — COMMENTED** [review]
Did a second review pass.

**Comment by MarchingCube** [comment]
Additionally since you still remember what new implementation does - a few class level comments on some of the internals wouldn't hurt (like `IValue`, property store types, etc.) since right now only you and I know how that stuff even works.

**Comment by grokys** [comment]
Thanks for the thorough review @MarchingCube! I think I've addressed all your feedback.

**Review by MarchingCube — APPROVED** [review]
LGTM!

**Pull request merged**
